coccinelle/with_mcxt.cocci

```diff
@@ -0,0 +1,29 @@
+@r1@
+expression mcxt;
+identifier oldmcxt;
+@@
+ {
+-MemoryContext oldmcxt;
+...
+-oldmcxt = MemoryContextSwitchTo(mcxt);
++TS_WITH_MEMORY_CONTEXT(mcxt, WRAP(
```

Comment:
It is intended to make coding easier as well and not only to avoid making mistakes. Not sure about if it is easy to check using a Coccinelle rule.

Reply:
> I don't really like this new wrapper, it's non-standard and feels a little cumbersome, so not sure we should require that it's used always. 

It is not required to be used always. It is possible to to use `ExplicitMemoryContext` if you need it for more lengthy code. The cocci rule only warns if you temporarily switch memory context in the same block and do not use this name, which is the common case.

Reply:
>What do you think about adding a Coccinelle rule that checks that the context was restored instead?

I think that this will not work. Quite often we switch memory context for other reasons and do not switch back in the same block, so trying to capture that will generate a lot of false negatives.

Reply:
> I think that this will not work. Quite often we switch memory context for other reasons and do not switch back in the same block, so trying to capture that will generate a lot of false negatives.

You mean, there are functions that don't switch the memory context back on return? That's unusual... I would say that's an error-prone interface.

You'll have to update these places in this PR, so that macro is not required, right?

Reply:
> Not sure about if it is easy to check using a Coccinelle rule.

We check a similar thing here, basically when one function call is not followed by another function call at some point: https://github.com/timescale/timescaledb/blob/main/coccinelle/heap_freetuple.cocci

Reply:
> > I think that this will not work. Quite often we switch memory context for other reasons and do not switch back in the same block, so trying to capture that will generate a lot of false negatives.
> 
> You mean, there are functions that don't switch the memory context back on return? That's unusual... I would say that's an error-prone interface.

Look for `MemoryContextSwitchTo` inside `tsl/src/nodes/decompress_chunk/exec.c`, `tsl/src/bgw_policy/job.c`, and possibly some more.

Reply:
As @akuzm proposed isn't much more simple just add a coccinelle script like
```
// find MemoryContextSwitchTo missing a context switch back
@ MemoryContextSwitch @
identifier oldctx;
position p;
@@

oldctx@p = MemoryContextSwitchTo(...);

@safelist@
expression oldctx;
position MemoryContextSwitch.p;
@@

oldctx@p = MemoryContextSwitchTo(...);
...
(
MemoryContextSwitchTo(oldctx)
)
@depends on !safelist@
expression oldctx;
position MemoryContextSwitch.p;
@@

+ /* MemoryContextSwitch missing context switch back */
  oldctx@p = MemoryContextSwitchTo(...);
```

It requires some more tweaks but I guess it should be enough!!!

Reply:
This particular script generates a false positive:

```
mats@fury:~/work/timescale/timescaledb+bugs$ spatch --very-quiet --sp-file test.cocci --dir src
diff -u -p a/home/mats/work/timescale/timescaledb+bugs/src/telemetry/telemetry.c b/home/mats/work/timescale/timescaledb+bugs/src/telemetry/telemetry.c
--- a/home/mats/work/timescale/timescaledb+bugs/src/telemetry/telemetry.c
+++ b/home/mats/work/timescale/timescaledb+bugs/src/telemetry/telemetry.c
@@ -395,6 +395,7 @@ add_errors_by_sqlerrcode(JsonbParseState
                add_errors_by_sqlerrcode_internal(parse_state,
                                                                                  TextDatumGetCString(record_jobtype),
                                                                                  sqlerrs_jsonb);
+               /* MemoryContextSwitch missing context switch back */
                old_context = MemoryContextSwitchTo(spi_context);
        }
 
@@ -522,6 +523,7 @@ add_job_stats_by_job_type(JsonbParseStat
                                                                        .total_duration_failures =
                                                                                DatumGetIntervalP(total_duration_failures) };
                add_job_stats_internal(parse_state, TextDatumGetCString(jobtype_datum), &stats);
+               /* MemoryContextSwitch missing context switch back */
                old_context = MemoryContextSwitchTo(spi_context);
        }
        res = SPI_finish();
```

I'll see if I can improve it though.